### ISSUE TITLE
How do you encorporate this into web frameworks?

### ISSUE THREAD

**Issue by yarbelk** [johnbywater/eventsourcing#53]
Specifically, I'm interested in how I would use the app in a django/uwsgi stack.  I'm not sure of the thread safety, or where the context manager should be set up.

my ideas are: 
- in a middleware
- in the views/api
- in a thread.local object.

**Comment by johnbywater** [comment]
It's a good question. You can do it in the wsgi file. If you have test cases, you can do it there too. Also if you have any workers (e.g. celery workers) you can do it when they are initialised. I suggest having a factory method that constructs and returns the application instance, and then calling that method from the wsgi file, so you have an eventsourcing application object as a module level variable, just like the wsgi application object.

The "rule" is that you only need one application instance per process. Otherwise, there will be more than one attempt to store each event, which won't work. I'm not going to use the word "singleton" because some get triggered by it, but it's just like imported Python modules or Python strings: you need only one instance per process.

The application object, in itself, is thread-safe because it is stateless. From experience, the difficulty some have had is the publish-subscribe functions it uses are at the level of the module, so that entity methods can publish events without having any other dependencies. So if you wanted (for some reason, e.g. in a test suite) to instantiate the application twice in the same process then you simply need to close it first: to unsubscribe handlers and to close database connections. Closing connections to services like databases and message brokers is only polite, and can save resources. Unsubscribing before resubscribing is essential to avoid trying to save each event more than once (which will immediately cause concurrency exceptions to be raised).

The context manager aspect is only a convenient way of closing things at the end. You can equally put the object in a try/except block and call close(). The 'with' context manager Python syntax is just a nice way to say that.

Since you will need to use the eventsourcing application object in the view methods, it's not good enough just to instantiate the application at the edge of your system: the views also need to have the object. So you could somehow pass the eventsourcing application object into the stack for each request, or you could code your factory method so that it can be called many times and each time will return the same object. If you do it like that, make sure that if the application has been closed, the factory will create a new instance and not just return a closed instance. You could call the factory function in each view method that needs it, or you could make it a property of a base view class (if you have one).

You could perhaps do it with some middleware, I didn't ever try doing it with middleware.

Hope that helps?

**Comment by johnbywater** [comment]
Also, if your process forks after it loads the code, you will want to open the application object in the child processes, after they have been forked. You also need to avoid race conditions, so just doing it lazily from multi-threaded requests often isn't good enough. Celery has a signal for it (```worker_process_init```). There are various things in Django which provide similar comfort. This can sometimes be a "tricky" area because sometimes Django developers don't really have a very good understanding of the processes in their services. Perhaps it would be better if the complexity was somehow encapsulated, so it is hidden and "just works". I guess we've got some more work to do with Django. I keep thinking it would be great to have something that doesn't break the Django admin view, but that's a different (much bigger) topic.

**Comment by yarbelkgazia** [comment]
This is a fantastic write up.  Thanks.  I've started a sphinx docs branch - i'm going to encorporate this into it.

**Comment by yarbelkgazia** [comment]
For example - using http://uwsgi-docs.readthedocs.io/en/latest/PythonDecorators.html#uwsgidecorators.postfork
and using a middle ware or borg/singleton/module initialization in the `wsgi.py`

```
# ... other stuff up here to setup settings, wrappers etc.

from django.core.wsgi import get_wsgi_application  # noqa
from uwsgidecorators import postfork
from eventsourcedapp import ev_app  # Import module - app isn't initiallized

application = get_wsgi_application()
app = get_wsgi_application()

@postfork
def setup_ev_app():
    ev_app.setup()
```

where `ev_app.setup()` does some of that module magic to actually setup the app and, and sets up the cleanup somehow.

**Comment by johnbywater** [comment]
Just to link things up, this reply is now in the docs here (has edit link): http://eventsourcing.readthedocs.io/en/latest/topics/user_guide/wsgi.html

Still missing your last comment about wsgi postfork decorator, and other things. And the "answer to question" style needs so it reads like normal documentation. So I won't close this issue just yet.

**Comment by johnbywater** [comment]
@yarbelk @yarbelkgazia I just rewrote the documentation about this, after giving it a bit more thought - see #68. Would be really great if you wanted to test this, to double check it isn't misleading or broken anything?

**Comment by yarbelk** [comment]
Absolutely,  I will check this in the morning as I'm now in a position to.  I was also planning on going over the documentation to check it as well (fresh pair of eyes and all that).

**Comment by johnbywater** [comment]
Thanks! Have been working on it some more, but things seem to have rounded out on this topic, so I've have merged everything I've done into the develop branch. If you didn't see it already, there's a Flask app that can run independently, and a wsgi file for it that runs with uWSGI, all tested. I backed off trying to do that in the user guide, which now just has some fragments that aren't really testable, and so aren't tested. I referred in the documentation to the working examples in the library, which are very simple but I feel demonstrate quite well how to place things. It's a really important thing to do, and it isn't immediately obvious, so I hope the documentation now illuminates the concerns adequately. I'm sure you'll let me know what you think ;-)

**Comment by renou** [comment]
@yarbelk we open sourced a django package using eventsourcing V1.x https://github.com/ApplauseOSS/djangoevents Maybe that helps you too :)

**Comment by johnbywater** [comment]
Thanks @renou! I'll mention it in the docs. We do need to update it though ;-)

**Comment by johnbywater** [comment]
@yarbelk is everything ok? didn't hear from you in a while, did I do something wrong? hope all's well with you

**Comment by johnbywater** [comment]
@renou have added a section linking to djangoevents, for good measure I put it in two places: http://eventsourcing.readthedocs.io/en/latest/topics/user_guide/schema.html#applause-djangoevents-project
http://eventsourcing.readthedocs.io/en/latest/topics/user_guide/wsgi.html#applause-djangoevents-project

**Comment by yarbelkgazia** [comment]
@johnbywater sorry, got sick and needed to catch up with work.

I had issues with the `@post_fork` method - and ended up with a `get_app` that memoized the initialized application.  Similar to how you did it in the docs.  Then in any method or function I needed the app in, I called the get_app function.

this ensures that the app is only initalized on its first use, which would be post fork.  I need to actually test this more for stability and re-grock how uwsgi forking behaviour impacts what is and isn't thread safe.  I knew this stuff a long while ago.

I also think this is impacted by which settings you pass to uwsgi... so:

## TL;DR

- Using something simillar to what you have in docs works in staging right now.
- `@post_fork` has issues i don't understand
- uwsgi is complicated

**Comment by renou** [comment]
@johnbywater 💥 breaking news! haha thank you! 😄

**Comment by johnbywater** [comment]
sorry to hear you have been unwell @yarbelkgazia :(

uwsgi is complicated! good to hear you got something on staging

**Comment by johnbywater** [comment]
have been rewriting this page, adding lots of links to things like Flask-SQLAlchemy package, http://eventsourcing.readthedocs.io/en/latest/index.html

also reworked the flaskapp.py module to use that package, which makes a more convincing example?

**Comment by johnbywater** [comment]
Going to close this issue because I think the new version of the docs, with the new example and links, covers this issue. ? Please reopen if you disagree :-)

http://eventsourcing.readthedocs.io/en/latest/topics/user_guide/wsgi.html
